### PR TITLE
mint community pool inflation in x/kavamint

### PR DESCRIPTION
**Pull request by pirtleshell**
Implements and tests `x/kavamint`'s `CommunityPoolInflation` parameter. Inflation is compounded for the number of seconds since the last block, and then coins are minted and transferred to the community pool (the module account of `x/community`)

Slightly modified staking reward implementation to ensure newly minted coins don't contribute the basis. (ie. if we mint coins and _then_ find the total supply, we'll be inflating more than expected because the total supply now includes the newly minted tokens.

### CODE REVIEW

File: x/kavamint/abci.go [hunk 1]
```diff
@@ -11,9 +11,12 @@ func BeginBlocker(ctx sdk.Context, k keeper.Keeper) {
 	if !found {
 		previousBlockTime = ctx.BlockTime()
 	}
+	// calculate totals before any minting is done to prevent new mints affecting the values
```
Comment (galxy25) [inline x/kavamint/abci.go:14]: does the ordering of the keepers in app.go ensure that x/mint's begin blocker is called before the staking modules?

File: x/kavamint/abci.go [hunk 2]
```diff
@@ -29,8 +32,20 @@ func BeginBlocker(ctx sdk.Context, k keeper.Keeper) {
 	}
 
 	// ------------- Community Pool -------------
-	// TODO: mint tokens for community pool
-	// TODO: send tokens to community pool
+	communityPoolInflation, err := k.AccumulateCommunityPoolInflation(ctx, totalSupply, previousBlockTime)
+	if err != nil {
+		panic(err)
```
Comment (galxy25) [inline x/kavamint/abci.go:37]: nit: might be useful to include the totalSupply and previousBlockTime in this error message?

**Review by galxy25 — APPROVED** [review]
could see updating the spec further to supplement the code samples with a written description of what the desired behavior for the mint module is and adding more edge case tests but otherwise lgtm

File: x/kavamint/abci.go [hunk 3]
```diff
@@ -11,9 +11,12 @@ func BeginBlocker(ctx sdk.Context, k keeper.Keeper) {
 	if !found {
 		previousBlockTime = ctx.BlockTime()
 	}
+	// calculate totals before any minting is done to prevent new mints affecting the values
+	totalBonded := k.TotalBondedTokens(ctx)
+	totalSupply := k.TotalSupply(ctx)
 
 	// ------------- Staking Rewards -------------
-	stakingRewardCoins, err := k.AccumulateStakingRewards(ctx, previousBlockTime)
+	stakingRewardCoins, err := k.AccumulateStakingRewards(ctx, totalBonded, previousBlockTime)
```
Comment (galxy25) [inline x/kavamint/abci.go:19]: an orthogonal concern to this pr, but I kinda think it makes since for the first begin blocker to calculate the time since the last block and pass that to all modules so that is consistent across modules

Comment (galxy25) [reply x/kavamint/abci.go:23]: (perhaps by setting it as value in the context...)

Comment (pirtleshell) [reply x/kavamint/abci.go]: i've refactored to consolidate `AccumulateStakingRewards` and `AccumulateCommunityPoolInflation` into a single `AccumulateInflation(ctx, rate, basis, secondsSinceLastMint)` method.

Now `secondsPassed` is calculated in the begin blocker and the individual inflation calculations look like:
```go
	stakingRewardCoins, err := k.AccumulateInflation(
		ctx, params.StakingRewardsApy, totalBonded, secondsPassed,
	)

	communityPoolInflation, err := k.AccumulateInflation(
		ctx, params.CommunityPoolInflation, totalSupply, secondsPassed,
	)
```

File: x/kavamint/abci.go (hunk 1 above)
Comment (pirtleshell) [reply x/kavamint/abci.go:21]: yes, the ordering of the begin blockers matters. the kavamint's begin blocker must be registered before distribution so that the fee pool has the newly minted tokens to distribute. indicated the requirement in a comment where the begin blocker is registered.

https://github.com/Kava-Labs/kava/blob/18c41c2ef15fa59af9a8d363302dd8ad03b73747/app/app.go#L783-L785
